packages/@eventual/aws-cdk/src/bucket-service.ts

```diff
@@ -0,0 +1,287 @@
+}
+
+export class BucketService<Service> {
+  public buckets: ServiceBuckets<Service>;
+  public bucketHandlers: ServiceBucketNotificationHandlers<Service>;
+
+  constructor(private props: BucketServiceProps<Service>) {
+    const bucketsScope = new Construct(props.serviceScope, "Buckets");
+
+    this.buckets = Object.fromEntries(
+      props.build.buckets.buckets.map((b) => [
+        b.name,
+        new Bucket(bucketsScope, {
+          bucket: b,
+          bucketService: this,
+          serviceProps: props,
+        }),
+      ])
+    ) as ServiceBuckets<Service>;
+
+    this.bucketHandlers = Object.values(
+      this.buckets as Record<string, Bucket>
+    ).reduce((handlers: Record<string, BucketNotificationHandler>, ent) => {
+      return {
+        ...handlers,
+        ...ent.handlers,
+      };
+    }, {}) as ServiceBucketNotificationHandlers<Service>;
+  }
+
+  public configureReadWriteBuckets(func: Function) {
+    this.addEnvs(func, ENV_NAMES.SERVICE_NAME, ENV_NAMES.BUCKET_OVERRIDES);
+    this.grantReadWriteBuckets(func);
+  }
+
+  public grantReadWriteBuckets(grantee: IGrantable) {
+    // find any bucket names that were provided by the service and not computed
+    const bucketNameOverrides = this.props.bucketOverrides
+      ? Object.values(
+          this.props.bucketOverrides as Record<string, BucketRuntimeOverrides>
+        )
+          .map((s) => s.bucketName)
+          .filter((s): s is string => !!s)
+      : [];
+
+    // grants the permission to start any task
+    grantee.grantPrincipal.addToPrincipalPolicy(
+      new PolicyStatement({
+        actions: ["s3:List*"],
+        resources: [
+          serviceBucketArn(
+            this.props.serviceName,
+            bucketServiceBucketSuffix("*"),
+            false
+          ),
+          ...bucketNameOverrides.map(formatBucketArn),
+        ],
+      })
+    );
+
+    grantee.grantPrincipal.addToPrincipalPolicy(
+      new PolicyStatement({
+        actions: [
+          "s3:List*",
+          "s3:GetObject*",
+          "s3:PutObject*",
+          "s3:DeleteObject*",
+        ],
+        resources: [
+          `${serviceBucketArn(
+            this.props.serviceName,
+            bucketServiceBucketSuffix("*"),
+            false
+          )}/*`,
+          ...bucketNameOverrides.map((s) => formatBucketArn(`${s}/*`)),
+        ],
+      })
+    );
+  }
+
+  private readonly ENV_MAPPINGS = {
+    [ENV_NAMES.SERVICE_NAME]: () => this.props.serviceName,
+    [ENV_NAMES.BUCKET_OVERRIDES]: () =>
+      Stack.of(this.props.serviceScope).toJsonString(
+        this.props.bucketOverrides
+      ),
+  } as const;
+
+  private addEnvs(func: Function, ...envs: (keyof typeof this.ENV_MAPPINGS)[]) {
+    envs.forEach((env) => func.addEnvironment(env, this.ENV_MAPPINGS[env]()));
+  }
+}
+
+interface BucketProps {
+  serviceProps: BucketServiceProps<any>;
+  bucketService: BucketService<any>;
+  bucket: BucketRuntime;
+}
+
+export class Bucket extends Construct {
```

Comment:
does this need to be exported?

Reply:
It gives direct access to the Bucket via the `service.buckets` member. However, it could be an interface instead of the construct I guess.